improver/calibration/reliability_calibration.py

```diff
@@ -0,0 +1,388 @@
+                             (forecast <= bin_max)).astype(int)
+            forecasts_probability_values = forecast * forecast_mask
+
+            observation_counts.append(observation_mask)
+            forecast_probabilities.append(forecasts_probability_values)
+            forecast_counts.append(forecast_mask)
+
+        reliability_table = np.stack([
+                                np.stack(observation_counts),
+                                np.stack(forecast_probabilities),
+                                np.stack(forecast_counts)])
+        return reliability_table.astype(np.float32)
+
+    @staticmethod
+    def _extract_truth(truths, forecast_slice, threshold_coord):
+        """
+        Extract a truth field that matches the threshold and time of the
+        forecast field.
+
+        Args:
+            truths (iris.cube.Cube):
+                The cube containing the thresholded truths used in calibration.
+            forecast_slice (iris.cube.Cube):
+                A slice over threshold and time of the forecast data.
+            threshold_coord (iris.coords.DimCoord):
+                The threshold coordinate that should be present on both input
+                cubes.
+        Returns:
+            iris.cube.Cube or None:
+                A truth cube that matches the time and threshold values of the
+                forecast slice, or None if no match is found.
+        """
+        threshold, = forecast_slice.coord(threshold_coord).points
+        threshold_name = threshold_coord.name()
+        threshold_constraint = {threshold_name: threshold}
+        time = forecast_slice.coord('time')
+
+        return truths.extract(iris.Constraint(
+            time=next(time.cells()).point, coord_values=threshold_constraint))
+
+    def process(self, historic_forecasts, truths):
+        """
+        Slice data over threshold and time coordinates to construct reliability
+        tables. These are summed over time to give a single table for each
+        threshold, constructed from all the provided historic forecasts and
+        truths.
+
+        Args:
+            historic_forecasts (iris.cube.Cube):
+                A cube containing the historical forecasts used in calibration.
+            truths (iris.cube.Cube):
+                A cube containing the thresholded gridded truths used in
+                calibration.
+        Returns:
+            iris.cube.CubeList:
+                A cubelist of reliability table cubes, one for each threshold
+                in the historic forecast cubes.
+        """
+        threshold_coord = find_threshold_coordinate(historic_forecasts)
+        time_coord = historic_forecasts.coord('time')
+
+        reliability_tables = iris.cube.CubeList()
+        for threshold_slice in historic_forecasts.slices_over(threshold_coord):
+
+            threshold_reliability = []
+            for time_slice in threshold_slice.slices_over(time_coord):
```

Comment:
Due to the use of the `cycle_hour` coordinate, I think that there's an assumption that the historic_forecasts will all be from the same `cycle_hour`. However, this isn't actually enforced, so currently the `cycle_hour` of the last forecast in the list provided would be used as the `cycle_hour`. It would be good to document this requirement and potentially enforce this.

Reply:
Done.